sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -169,7 +170,7 @@ public PBegin begin() {
    * @see #apply(String, PTransform)
    */
   public <OutputT extends POutput> OutputT apply(
-      PTransform<? super PBegin, OutputT> root) {
+      PTransform<? super PBegin, OutputT> root) throws ValidationException {
```

Comment:
+1, the current change is overly invasive. I would recommend to also do it in 3 separate commits (ok to do in 1 PR though) - one for Source.validate, one for PTransform.validate, one for Pipeline.validate.

Also, I think this requires more design/discussion than just adding a throws to the signature: we also need guidelines. E.g. are we going to _require_ that all errors be reported via ValidationException? I hope not - e.g. I'd like to still use Preconditions.checkArgument() for the easy stuff. If not - then which errors should be wrapped in ValidationException and which shouldn't?